#!/bin/bash 

#make the directory for the assessment 
mkdir ngs_assessment
mkdir ngs_assessment/dnaseq

#creating the directory tree
cd ngs_assessment/dnaseq
mkdir data meta results logs scripts

# moving to the data directory and creating folders that will hold the data moving forward
cd ~/ngs_assessment/dnaseq/data
mkdir untrimmed_fastq
mkdir trimmed_fastq

#downloading the input files and moving them to their locations
wget https://s3-eu-west-1.amazonaws.com/workshopdata2017/NGS0001.R1.fastq.qz

wget https://s3-eu-west-1.amazonaws.com/workshopdata2017/NGS0001.R2.fastq.qz

wget https://s3-eu-west-1.amazonaws.com/workshopdata2017/annotation.bed

mv *fastq.qz ~/ngs_assessment/dnaseq/data/untrimmed_fastq
mv annotation.bed ~/ngs_assessment/dnaseq/data

#downloading the hg19 reference and moving it 
wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz

mv hg19.fa.gz ~/ngs_assessment/dnaseq/data

# enter directory with raw FASTQs
cd ~/ngs_assessment/dnaseq/data/untrimmed_fastq

#uncompress the fastq files
zcat NGS0001.R1.fastq.qz > NGS0001.R1.fastq
zcat NGS0001.R2.fastq.qz > NGS0001.R2.fastq

#giving the fastqc files as input
fastqc -t 4 *.fastq.qz

#creating a sub-directory for the fastqc untrimmed reads in the results directory
mkdir ~/ngs_assessment/dnaseq/results/fastqc_untrimmed_reads
mv *fastqz* ~/ngs_assessment/dnaseq/results/fastqc_untrimmed_reads

#looking at the files generated by FastQC
ls -lh ~/ngs_assessment/dnaseq/results/fastqc_untrimmed_reads

#unzipping the .zip files that form the output of FastQC
for zip in *.zip
do
unzip $zip
done

#change directories to untrimmed fastq location
cd ~/ngs_assessment/dnaseq/data/untrimmed_fastq

#running trimmomatic
trimmomatic PE -threads 4  -phred33  /home/ubuntu/ngs_assessment/dnaseq/data/untrimmed_fastq/NGS0001.R1.fastq /home/ubuntu/ngs_assessment/dnaseq/data/untrimmed_fastq/NGS0001.R2.fastq -baseout /home/ubuntu/ngs_assessment/dnaseq/data/trimmed_fastq/NGS0001_trimmed_R ILLUMINACLIP:/home/ubuntu/anaconda3/pkgs/trimmomatic-0.39-hdfd78af_2/share/trimmomatic/adapters/NexteraPE-PE.fa:2:30:10 TRAILING:25 MINLEN:50

#change directories to trimmed fastq location
cd ~/ngs_assessment/dnaseq/data/trimmed_fastq

#performing quality assessment of paired trimmed sequencing data
fastqc NGS0001_trimmed_R_1P
fastqc NGS0001_trimmed_R_2P

#making a location for the results
mkdir ~/ngs_assessment/dnaseq/results/fastqc_trimmed_reads
mv *fastqc* ~/ngs_assessment/dnaseq/results/fastqc_trimmed_reads

#looking at the files generated from fastqc
ls -lh ~/ngs_assessment/dnaseq/results/fastqc_trimmed_reads

#unzipping the zip files that form a part of the results from fastqc
for zip in *.zip
do
unzip $zip
done

#making folders for the reference index and its index file, then running bwa to get the index files
mkdir -p ~/ngs_assessment/dnaseq/data/reference
mv ~/ngs_assessment/dnaseq/data/hg19.fa.gz ~/ngs_assessment/dnaseq/data/reference
bwa index ~/ngs_assessment/dnaseq/data/reference/hg19.fa.gz
ls ~/ngs_assessment/dnaseq/data/reference

#running bwa mem with read group info and including it in the bam file
mkdir ~/ngs_assessment/dnaseq/data/aligned_data

bwa mem -t 4 -v 1 -R '@RG\tID:HWI-D0011.50.H7AP8ADXX.1.NGS0001\tSM:NGS0001\tPL:ILLUMINA\tLB:nextera-ngs0001-blood\tDT:2017-02-23\tPU:HWI-D00119' -I 250,50  ~/ngs_assessment/dnaseq/data/reference/hg19.fa.gz ~/ngs_assessment/dnaseq/data/trimmed_fastq/NGS0001_trimmed_R_1P ~/ngs_assessment/dnaseq/data/trimmed_fastq/NGS0001_trimmed_R_2P > ~/ngs_assessment/dnaseq/data/aligned_data/NGS0001.sam

#change directories to aligned data 
cd ~/ngs_assessment/dnaseq/data/aligned_data

#using samtools we convert the sam to bam file, sort it and generate an index 
samtools view -h -b NGS0001.sam > NGS0001.bam
samtools sort NGS0001.bam > NGS0001_sorted.bam
samtools index NGS0001_sorted.bam
ls

#MarkDuplicates
picard MarkDuplicates I=NGS0001_sorted.bam O=NGS0001_sorted_marked.bam M=marked_dup_metrics.txt
samtools index NGS0001_sorted_marked.bam

#Filter BAM on mapping quality and bitwise flags using samtools
samtools view -F 1796  -q 20 -o NGS0001_sorted_filtered.bam  NGS0001_sorted_marked.bam
samtools index NGS0001_sorted_filtered.bam

#generating standard alignment statistics
#1 flagstat
samtools flagstat NGS0001_sorted_filtered.bam
#2 idxstats
samtools idxstats NGS0001_sorted_filtered.bam
#3 insert size (picard)
picard CollectInsertSizeMetrics I=NGS0001_sorted_filtered.bam O=marked_and_filtered_insertsize.stats H=marked_and_filtered_insertsize.pdf
#4 changing the location of annotation.bed so bedtools can be run to get depth of coverage
mv ~/ngs_assessment/dnaseq/data/annotation.bed ~/ngs_assessment/dnaseq/data/aligned_data
bedtools coverage -a NGS0001_sorted_filtered.bam -b annotation.bed > depthofcoverage.txt

#unzip the hg19 indexed file
zcat ~/ngs_assessment/dnaseq/data/reference/hg19.fa.gz > ~/ngs_assessment/dnaseq/data/reference/hg19.fa
samtools faidx ~/ngs_assessment/dnaseq/data/reference/hg19.fa
#freebayes
freebayes --bam ~/ngs_assessment/dnaseq/data/aligned_data/NGS0001_sorted_filtered.bam --fasta-reference ~/ngs_assessment/dnaseq/data/reference/hg19.fa --vcf ~/ngs_assessment/dnaseq/results/NGS0001.vcf
bgzip ~/ngs_assessment/dnaseq/results/NGS0001.vcf
tabix -p vcf ~/ngs_assessment/dnaseq/results/NGS0001.vcf.gz


#filtering the vcf
vcffilter -f "QUAL > 1 & QUAL / AO > 10 & SAF > 0 & SAR > 0 & RPR > 1 & RPL > 1" \
        ~/ngs_assessment/dnaseq/results/NGS0001.vcf.gz > ~/ngs_assessment/dnaseq/results/NGS0001_filtered.vcf
#intersect with the given annotation file to restrict analysis to regions of interest in bed file
bedtools intersect -header -wa -a ~/ngs_assessment/dnaseq/results/NGS0001_filtered.vcf -b ~/ngs_assessment/dnaseq/data/aligned_data/annotation.bed \
        > ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.vcf

bgzip ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.vcf
tabix -p vcf ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.vcf.gz

#snpEff
cd ~/ngs_assessment/dnaseq/data
snpEff -Xmx8G -v GRCh37.75  ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.vcf.gz > ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.snpeff.vcf -csvStats ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.snpeff.csv -s ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.snpeff.html


#downloading annovar and setting up the necessities 
cd ~/ngs_assessment/dnaseq/data
tar -zxvf annovar.latest.tar.gz
#change directory to where annovar is downloaded
cd annovar
./annotate_variation.pl -buildver hg19 -downdb -webfrom annovar knownGene humandb/
./annotate_variation.pl -buildver hg19 -downdb -webfrom annovar refGene humandb/
./annotate_variation.pl -buildver hg19 -downdb -webfrom annovar ensGene humandb/
./annotate_variation.pl -buildver hg19 -downdb -webfrom annovar clinvar_20180603 humandb/
./annotate_variation.pl -buildver hg19 -downdb -webfrom annovar exac03 humandb/
./annotate_variation.pl -buildver hg19 -downdb -webfrom annovar dbnsfp31a_interpro humandb/

./convert2annovar.pl -format vcf4 ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.vcf.gz > ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.avinput

#creating an excel output table 
./table_annovar.pl ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated.avinput humandb/ -buildver hg19    -out ~/ngs_assessment/dnaseq/results/NGS0001_filtered_annotated -remove    -protocol refGene,ensGene,clinvar_20180603,exac03,dbnsfp31a_interpro -operation g,g,f,f,f -otherinfo -nastring . -csvout
